### ISSUE TITLE
Visualization

### ISSUE THREAD

**Issue by hz658832** [avishkarsaha/translating-images-into-maps#4]
Dear @avishkarsaha,

thanks for the awesome work and sharing the code!

Is it possible that you also share the visualization code? There may be some files missed under ./src, such as "src.visualization".

I am highly interested in the performance of large-scale environments. 

Thank you!

Bests,

Haoming

**Comment by caifazhou** [comment]
The visualization is missing in current version. Hope it can be added soon for easir reproducing

**Comment by Handsomer1999** [comment]
The visualization is missing

**Comment by MrChangQ** [comment]
> Dear @avishkarsaha,
> 
> thanks for the awesome work and sharing the code!
> 
> Is it possible that you also share the visualization code? There may be some files missed under ./src, such as "src.visualization".
> 
> I am highly interested in the performance of large-scale environments.
> 
> Thank you!
> 
> Bests,
> 
> Haoming

you can try this function which is writed by myself to show the result.


def visualize_score(scores, heatmaps, grid, image, iou, num_classes):
    # Condese scores and ground truths to single map
    class_idx = torch.arange(len(scores)) + 1
    logits = scores.clone().cpu() * class_idx.view(-1, 1, 1)
    logits, _ = logits.max(dim=0)
    scores = (scores.detach().clone().cpu() > 0.5).float() * class_idx.view(-1, 1, 1)
    cls_idx = scores.clone()
    cls_idx = cls_idx.argmax(dim=0)
    cls_idx = cls_idx.numpy() * 20
    cls_idx = cv2.applyColorMap(cv2.convertScaleAbs(cls_idx,alpha=1),cv2.COLORMAP_JET)
    # print("cls_idx: ",cls_idx)
    # print("scores: ",scores )
    scores, _ = scores.max(dim=0)
    heatmaps = (heatmaps.detach().clone().cpu() > 0.5).float() * class_idx.view(
        -1, 1, 1
    )
    heatmaps, _ = heatmaps.max(dim=0)

    # Visualize score
    fig = plt.figure(num="score", figsize=(8, 6))
    fig.clear()

    gs = mpl.gridspec.GridSpec(2, 3, figure=fig)
    ax1 = fig.add_subplot(gs[0, :])
    ax2 = fig.add_subplot(gs[1, 0])
    ax3 = fig.add_subplot(gs[1:, 1])
    ax4 = fig.add_subplot(gs[1:, 2])

    image = ax1.imshow(image)
    ax1.grid(which="both")
    ax2 = ax2.imshow(cls_idx)
    # print("logits : ",logits.shape,logits)
    # print("scores : ",scores)
    # scores_threshold = np.where(scores > 0.5, scores, 0)
    ax3 = ax3.imshow(scores)
    ax4 = ax4.imshow(heatmaps)
    # plt.show()
    # src.visualization.encoded.vis_score_raw(logits, grid, cmap="magma", ax=ax2)
    # src.vis_score(scores, grid, cmap="magma", ax=ax3, num_classes=num_classes)
    # src.vis_score(heatmaps, grid, cmap="magma", ax=ax4, num_classes=num_classes)

    grid = grid.cpu().detach().numpy()
    yrange = np.arange(grid[:, 0].max(), step=5)
    xrange = np.arange(start=grid[0, :].min(), stop=grid[0, :].max(), step=5)
    ymin, ymax = 0, grid[:, 0].max()
    xmin, xmax = grid[0, :].min(), grid[0, :].max()

    ax2 = plt.vlines(xrange, ymin, ymax, color="white", linewidth=0.5)
    ax2 = plt.hlines(yrange, xmin, xmax, color="white", linewidth=0.5)
    ax3 = plt.vlines(xrange, ymin, ymax, color="white", linewidth=0.5)
    ax3 = plt.hlines(yrange, xmin, xmax, color="white", linewidth=0.5)
    ax4 = plt.vlines(xrange, ymin, ymax, color="white", linewidth=0.5)
    ax4 = plt.hlines(yrange, xmin, xmax, color="white", linewidth=0.5)

    ax1.set_title("Input image", size=11)
    # ax2.set_title("Model output logits", size=11)
    # ax3.set_title("Model prediction = logits" + r"$ > 0.5$", size=11)
    # ax4.set_title("Ground truth", size=11)

    # plt.suptitle(
    #     "IoU : {:.2f}".format(iou), size=14,
    # )

    gs.tight_layout(fig)
    gs.update(top=0.9)

    return fig

**Comment by wang-TJ-20** [comment]
@WilliamLucifer I use this code, but the output is this image 
![image](https://user-images.githubusercontent.com/99472631/184071586-4d27e266-2d28-4e8c-b05e-c2af5587d25a.png)

**Comment by WilliamLucifer** [comment]
> @WilliamLucifer thanks for the guidance with the visualization code! Quick question:
> 
> For your ground truth visual in the fourth subplot produced from your code (ax4), are you getting a strange small white grid in the top left corner of the plot? I'm not sure where it is coming from, but the ground truth image otherwise looks fine.
> 
> <img alt="Screen Shot 2022-10-06 at 1 04 00 AM" width="572" src="https://user-images.githubusercontent.com/74636624/194219043-b3d79c0b-4d33-452e-9e32-247f78b3f6bc.png">
> 
> I looked at all of the heatmaps individually with color, and none of them had that small grid in the top right. Any ideas?

**Comment by WilliamLucifer** [comment]
First,I think you can try to annotate all your codes about the function of grid like "plt.grid".Then,you maybe can get a result which you want to get.Finally,you can check why the function of grid does not work like the error location or error size.
I hope it can help you.

**Comment by kevinvlowell** [comment]
> > @WilliamLucifer thanks for the guidance with the visualization code! Quick question:
> > For your ground truth visual in the fourth subplot produced from your code (ax4), are you getting a strange small white grid in the top left corner of the plot? I'm not sure where it is coming from, but the ground truth image otherwise looks fine.
> > <img alt="Screen Shot 2022-10-06 at 1 04 00 AM" width="572" src="https://user-images.githubusercontent.com/74636624/194219043-b3d79c0b-4d33-452e-9e32-247f78b3f6bc.png">
> > I looked at all of the heatmaps individually with color, and none of them had that small grid in the top right. Any ideas?
> 
> First,I think you can try to annotate all your codes about the function of grid like "plt.grid".Then,you maybe can get a result which you want to get.Finally,you can check why the function of grid does not work like the error location or error size. I hope it can help you.

Thanks so much for your reply, William! I've realized that the small grid is just placed in whichever subplot is the last subplot shown, meaning its not specific to heatmaps or any other plotted data. I'm just confused what the purpose of the grid is/what it should look like, so I'm wondering if that will mess up my training results or something. I used the code exactly as you provided, so I'm wondering if grid production via the training process is not working correctly.

Would you mind sharing what your visualization looked like for reference? I'd greatly appreciate it!

**Comment by xuxian-0205** [comment]
> > Dear @avishkarsaha,
> > thanks for the awesome work and sharing the code!
> > Is it possible that you also share the visualization code? There may be some files missed under ./src, such as "src.visualization".
> > I am highly interested in the performance of large-scale environments.
> > Thank you!
> > Bests,
> > Haoming
> 
> you can try this function which is writed by myself to show the result.
> 
> def visualize_score(scores, heatmaps, grid, image, iou, num_classes): # Condese scores and ground truths to single map class_idx = torch.arange(len(scores)) + 1 logits = scores.clone().cpu() * class_idx.view(-1, 1, 1) logits, _ = logits.max(dim=0) scores = (scores.detach().clone().cpu() > 0.5).float() * class_idx.view(-1, 1, 1) cls_idx = scores.clone() cls_idx = cls_idx.argmax(dim=0) cls_idx = cls_idx.numpy() * 20 cls_idx = cv2.applyColorMap(cv2.convertScaleAbs(cls_idx,alpha=1),cv2.COLORMAP_JET) # print("cls_idx: ",cls_idx) # print("scores: ",scores ) scores, _ = scores.max(dim=0) heatmaps = (heatmaps.detach().clone().cpu() > 0.5).float() * class_idx.view( -1, 1, 1 ) heatmaps, _ = heatmaps.max(dim=0)
> 
> ```
> # Visualize score
> fig = plt.figure(num="score", figsize=(8, 6))
> fig.clear()
> 
> gs = mpl.gridspec.GridSpec(2, 3, figure=fig)
> ax1 = fig.add_subplot(gs[0, :])
> ax2 = fig.add_subplot(gs[1, 0])
> ax3 = fig.add_subplot(gs[1:, 1])
> ax4 = fig.add_subplot(gs[1:, 2])
> 
> image = ax1.imshow(image)
> ax1.grid(which="both")
> ax2 = ax2.imshow(cls_idx)
> # print("logits : ",logits.shape,logits)
> # print("scores : ",scores)
> # scores_threshold = np.where(scores > 0.5, scores, 0)
> ax3 = ax3.imshow(scores)
> ax4 = ax4.imshow(heatmaps)
> # plt.show()
> # src.visualization.encoded.vis_score_raw(logits, grid, cmap="magma", ax=ax2)
> # src.vis_score(scores, grid, cmap="magma", ax=ax3, num_classes=num_classes)
> # src.vis_score(heatmaps, grid, cmap="magma", ax=ax4, num_classes=num_classes)
> 
> grid = grid.cpu().detach().numpy()
> yrange = np.arange(grid[:, 0].max(), step=5)
> xrange = np.arange(start=grid[0, :].min(), stop=grid[0, :].max(), step=5)
> ymin, ymax = 0, grid[:, 0].max()
> xmin, xmax = grid[0, :].min(), grid[0, :].max()
> 
> ax2 = plt.vlines(xrange, ymin, ymax, color="white", linewidth=0.5)
> ax2 = plt.hlines(yrange, xmin, xmax, color="white", linewidth=0.5)
> ax3 = plt.vlines(xrange, ymin, ymax, color="white", linewidth=0.5)
> ax3 = plt.hlines(yrange, xmin, xmax, color="white", linewidth=0.5)
> ax4 = plt.vlines(xrange, ymin, ymax, color="white", linewidth=0.5)
> ax4 = plt.hlines(yrange, xmin, xmax, color="white", linewidth=0.5)
> 
> ax1.set_title("Input image", size=11)
> # ax2.set_title("Model output logits", size=11)
> # ax3.set_title("Model prediction = logits" + r"$ > 0.5$", size=11)
> # ax4.set_title("Ground truth", size=11)
> 
> # plt.suptitle(
> #     "IoU : {:.2f}".format(iou), size=14,
> # )
> 
> gs.tight_layout(fig)
> gs.update(top=0.9)
> 
> return fig
> ```
Could you give some samples about how to use this function?I don't know where this should be in.

**Comment by hz658832** [comment]
Dear @avishkarsaha,
Thanks for the solution, and sorry I haven't responded for a long time...
I believe this problem is solved. If someone still has a problem, please start with a new issue. 
Thanks!
Haoming

**Comment by MQC1111** [comment]
> > > Dear @avishkarsaha,
> > > thanks for the awesome work and sharing the code!
> > > Is it possible that you also share the visualization code? There may be some files missed under ./src, such as "src.visualization".
> > > I am highly interested in the performance of large-scale environments.
> > > Thank you!
> > > Bests,
> > > Haoming
> > 
> > 
> > you can try this function which is writed by myself to show the result.
> > def visualize_score(scores, heatmaps, grid, image, iou, num_classes): # Condese scores and ground truths to single map class_idx = torch.arange(len(scores)) + 1 logits = scores.clone().cpu() * class_idx.view(-1, 1, 1) logits, _ = logits.max(dim=0) scores = (scores.detach().clone().cpu() > 0.5).float() * class_idx.view(-1, 1, 1) cls_idx = scores.clone() cls_idx = cls_idx.argmax(dim=0) cls_idx = cls_idx.numpy() * 20 cls_idx = cv2.applyColorMap(cv2.convertScaleAbs(cls_idx,alpha=1),cv2.COLORMAP_JET) # print("cls_idx: ",cls_idx) # print("scores: ",scores ) scores, _ = scores.max(dim=0) heatmaps = (heatmaps.detach().clone().cpu() > 0.5).float() * class_idx.view( -1, 1, 1 ) heatmaps, _ = heatmaps.max(dim=0)
> > ```
> > # Visualize score
> > fig = plt.figure(num="score", figsize=(8, 6))
> > fig.clear()
> > 
> > gs = mpl.gridspec.GridSpec(2, 3, figure=fig)
> > ax1 = fig.add_subplot(gs[0, :])
> > ax2 = fig.add_subplot(gs[1, 0])
> > ax3 = fig.add_subplot(gs[1:, 1])
> > ax4 = fig.add_subplot(gs[1:, 2])
> > 
> > image = ax1.imshow(image)
> > ax1.grid(which="both")
> > ax2 = ax2.imshow(cls_idx)
> > # print("logits : ",logits.shape,logits)
> > # print("scores : ",scores)
> > # scores_threshold = np.where(scores > 0.5, scores, 0)
> > ax3 = ax3.imshow(scores)
> > ax4 = ax4.imshow(heatmaps)
> > # plt.show()
> > # src.visualization.encoded.vis_score_raw(logits, grid, cmap="magma", ax=ax2)
> > # src.vis_score(scores, grid, cmap="magma", ax=ax3, num_classes=num_classes)
> > # src.vis_score(heatmaps, grid, cmap="magma", ax=ax4, num_classes=num_classes)
> > 
> > grid = grid.cpu().detach().numpy()
> > yrange = np.arange(grid[:, 0].max(), step=5)
> > xrange = np.arange(start=grid[0, :].min(), stop=grid[0, :].max(), step=5)
> > ymin, ymax = 0, grid[:, 0].max()
> > xmin, xmax = grid[0, :].min(), grid[0, :].max()
> > 
> > ax2 = plt.vlines(xrange, ymin, ymax, color="white", linewidth=0.5)
> > ax2 = plt.hlines(yrange, xmin, xmax, color="white", linewidth=0.5)
> > ax3 = plt.vlines(xrange, ymin, ymax, color="white", linewidth=0.5)
> > ax3 = plt.hlines(yrange, xmin, xmax, color="white", linewidth=0.5)
> > ax4 = plt.vlines(xrange, ymin, ymax, color="white", linewidth=0.5)
> > ax4 = plt.hlines(yrange, xmin, xmax, color="white", linewidth=0.5)
> > 
> > ax1.set_title("Input image", size=11)
> > # ax2.set_title("Model output logits", size=11)
> > # ax3.set_title("Model prediction = logits" + r"$ > 0.5$", size=11)
> > # ax4.set_title("Ground truth", size=11)
> > 
> > # plt.suptitle(
> > #     "IoU : {:.2f}".format(iou), size=14,
> > # )
> > 
> > gs.tight_layout(fig)
> > gs.update(top=0.9)
> > 
> > return fig
> > ```
> 
> Could you give some samples about how to use this function?I don't know where this should be in.

You have successfully implemented visualization. How did you do it? Can you share it? Thanks a lot !!!
